Make it possible to acquire OIDC SecurityIdentity after HTTP request has completed (#36631)
quarkusio/quarkus: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc

Note that tenant specific configurations can also be disabled, for example: `quarkus.oidc.tenant-a.tenant-enabled=false`.

== Authentication outside of the HTTP request context

Sometimes it is practical to be able to create `SecurityIdentity` for given token when there is no active request context.
The `quarkus-oidc` extension uses its own `OidcIdentityProvider` to convert a token to a `SecurityIdentity` instance.
When the provider is used during an HTTP request, the tenant configuration can be resolved as described in the <<static-tenant-resolution>> section.
However, when there is no active HTTP request, you need to select tenant explicitly with the `io.quarkus.oidc.TenantFeature` qualifier like in the example below:

[source,java]
----
package org.acme.quickstart.oidc;

import io.quarkus.oidc.TenantFeature;
import io.quarkus.oidc.runtime.OidcIdentityProvider;
import io.quarkus.runtime.StartupEvent;
import io.quarkus.security.identity.SecurityIdentity;
import io.smallrye.mutiny.Uni;
import jakarta.enterprise.event.Observes;

public class StartupService {

>     void init(@Observes StartupEvent event, @TenantFeature("tenantId") OidcIdentityProvider identityProvider) {

Reviewer: It looks fine, but I'd like to propose to have an example showing what seems to be a more typical case, the request comes in, the token is passed to some other thread, the request has gone, and the new runner verifies the token

Reviewer: Please keep this example, but show the proposed example first

Author: Charm of this example is that everyone will understand it straight away, while some threading stuff is going to be complicated, they will need to think and hope they figure what I meant. 

**Please don't get me wrong, I'll do it if you are convinced of this**, but before let me to provide few points to which you can react:

- it is very close to scheduled tasks https://quarkus.io/guides/scheduler I suppose user will use service account where possible, but #24069 explicitly mentioned scheduled methods
- > more typical case, the request comes in, the token is passed to some other thread, the request has gone, and the new runner verifies the token

if the only problem is that HTTP request passed, can you leverage context propagation instead of this? I take it there, `RoutingContext` will still be stored in the CDI request context, so it seems to me that I would be suggesting something that users shouldn't do.

- I think the point here is to be able to use `OidcIdentityProvider` independently on HTTP request, do people really store old access token in real world situation from ended HTTP request and use it without the original CDI request context? I don't know, maybe there were users that already mentioned this?

Author: There is one else thing @sberyozkin  and please push back, I'm just asking: why would users really store old tokens instead of injecting identity when request is active and use it in the new thread when necessary?

Author: sorry @sberyozkin I didn't mention following comment `Please keep this example, but show the proposed example first`. I'll do it..

Reviewer: @michalvavrik Np, @knutwannheden may provide some context, and I also saw something similar asked in Discussions, #27754, etc

Reviewer: Thinking more about it, `getAdminToken` - how would anyone implement it, using `OidcTestClient` ? The use case is to verify the access token which was left by the completed request, tokens acquired by clients in scope of the current endpoint would not be targeting this endpoint, they are meant to be propagated to downstream services (Keycloak, or some microservices). It is an interesting example, but IMHO it may confuse users as to the main goal of this enhancement

Reviewer: >messing with threads is a bad idea

Why ? Users may have different use cases, though indeed, as I just commented above, if the token is passed out of band, then it is just a case of the manual verification, no threads are involved.

Author: Use case is legit, but documentation should be simple if possible. I'm not sure what is level of users knowledge on this topic and if it is possible to show example without using different executor, I'd do it.

Reviewer: @michalvavrik So may be lets just show a manual verification of the token passed for example with the custom header, or as a JSON property ?

Author: done. I'll close this thread as I addressed suggestion and we many thread comments opened. Please re-open if you are not satisfied or create a new one.

Note that tenant specific configurations can also be disabled, for example: `quarkus.oidc.tenant-a.tenant-enabled=false`.

== Authentication outside of the HTTP request context

> 

Reviewer: Great example

Author: Thank you!

When the provider is used during an HTTP request, the tenant configuration can be resolved as described in the <<static-tenant-resolution>> section.
However, when there is no active HTTP request, you need to select tenant explicitly with the `io.quarkus.oidc.TenantFeature` qualifier.
One situation when you need to select the tenant configuration this way is when you are processing messages with the xref:vertx.adoc#event-bus[Vert.x event bus].
In the example below, the 'product-order' message is consumed within different CDI request context, therefore injected `SecurityIdentity` would be anonymous.

[source,java]
----
package org.acme.quickstart.oidc;

import jakarta.annotation.security.RolesAllowed;
import jakarta.inject.Inject;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import org.eclipse.microprofile.jwt.JsonWebToken;
import io.vertx.core.eventbus.EventBus;

@Path("order")
public class OrderResource {

    @Inject
    EventBus eventBus;

    @Inject
    JsonWebToken accessToken;

    @RolesAllowed("customer")
    @POST
    public void order(String product) {

>         eventBus.publish("product-order", new Product(product, 1, accessToken.getRawToken()));

Reviewer: Sorry, example is indeed good, but here we already have a verified token because the method is secure. So it does not describe the use case after all. I guess you need to drop `@RolesAllowed` and disable the proactive authentication, or perhaps, possibly in addition to the updated example, show a token passed as a custom header param, in which case you don't even need to deal with threads

Author: I'll just do it.

Author: I must admit my opinion slightly differ, IMO it's still fine that token has already been verified because you still need a way to convert token to the identity. Anyway, I believe I completely addressed this comment and I also added part about disabled proactive auth as you suggester.

    }
}
----

[source,java]
----
package org.acme.quickstart.oidc;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import io.quarkus.oidc.TenantFeature;
import io.quarkus.oidc.runtime.OidcIdentityProvider;
import io.quarkus.security.identity.SecurityIdentity;
import io.quarkus.vertx.ConsumeEvent;
import io.smallrye.common.annotation.Blocking;

@ApplicationScoped
public class OrderService {

    @TenantFeature("tenantId")
    @Inject

>     OidcIdentityProvider identityProvider;

Reviewer: It looks good but I'm concerned of making `OidcIdentityProvider` as the user centric API point. `import io.quarkus.oidc.runtime.*` package keeps API, classes, which I expect users interact with.

I've been thinking of a dedicated API interface `TenantIdentyProvider` for example, and its `default` implementation would call the OidcIdentityProvider directly, or may be it will get a list of all identity providers, and get those which accept token requests to verify it, or users would do something else with custom implementations.

IMHO it will give us more flexibility, what do you think ?

Reviewer: @michalvavrik If we don't do it now then it may end up going directly to 3.6.0, as you may be overloaded as you already are :-).  How about:
```
public interface TenantIdentyProvider {
  Uni<SecurityIdentity> authenticate(AccessTokenCredential token);
}
```
for now ? OidcIdentityProvider makes some decision depending on whether it is an access or id token; we can add another method(s) later on depending on the requirement

Author: Fine with me, I'd still expect this to work fine for id token once there is use case.

Author: Done. Please feel free to suggest changes.

Note that tenant specific configurations can also be disabled, for example: `quarkus.oidc.tenant-a.tenant-enabled=false`.

> == Authentication outside of the HTTP request context

Reviewer: ```suggestion
== Authentication after HTTP request has completed
```

> Sometimes it is practical to be able to create `SecurityIdentity` for given token when there is no active request context.

Reviewer: ```suggestion
Sometimes, `SecurityIdentity` for a given token must be created when there is no active HTTP request context.
```

The `quarkus-oidc` extension provides `io.quarkus.oidc.TenantIdentityProvider` to convert a token to a `SecurityIdentity` instance.

> When the provider is used during an HTTP request, the tenant configuration can be resolved as described in the <<static-tenant-resolution>> section.

Reviewer: I'd like to propose adding the advice from lines 851, 852, as a NOTE when we show an example, IMHO this feature is really not about multi-tenancy, I believe the use case we discussed earlier, with only the default tenant would be typical, but indeed, we'd explain, that if needed, users can select a given tenant configuration the way you describe in the lines 850, 851. 
Adding a note that it is currently not supported for the dynamic tenant resolution is worth it too IMHO

Author: > Adding a note that it is currently not supported for the dynamic tenant resolution is worth it too IMHO

There is a warning bit later in the section.

However, when there is no active HTTP request, you need to select tenant explicitly with the `io.quarkus.oidc.TenantFeature` qualifier.

> One situation when you need to select the tenant configuration this way is when you are processing messages with the xref:vertx.adoc#event-bus[Vert.x event bus].

Reviewer: ```suggestion
For example, one situation when you need to verify the token after HTTP request has completed is when you are processing messages with the xref:vertx.adoc#event-bus[Vert.x event bus].
```

Reviewer: Suggesting not to focus the attention on the tenant configuration but on the main feature

> In the example below, the 'product-order' message is consumed within different CDI request context, therefore injected `SecurityIdentity` would be anonymous.

Reviewer: ```suggestion
In the example below, the 'product-order' message is consumed within different CDI request context, therefore an injected `SecurityIdentity` would not correctly represent the verified identity and be anonymous.
```

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import io.quarkus.oidc.AccessTokenCredential;
import io.quarkus.oidc.TenantFeature;
import io.quarkus.oidc.TenantIdentityProvider;
import io.quarkus.security.identity.SecurityIdentity;
import io.quarkus.vertx.ConsumeEvent;
import io.smallrye.common.annotation.Blocking;

@ApplicationScoped
public class OrderService {

    @TenantFeature("tenantId")
    @Inject
    TenantIdentityProvider identityProvider;

    @Inject
    TenantIdentityProvider defaultIdentityProvider; <1>

    @Blocking
    @ConsumeEvent("product-order")
    void processOrder(Product product) {
        String rawToken = product.customerAccessToken;
        AccessTokenCredential token = new AccessTokenCredential(rawToken);
        SecurityIdentity = identityProvider.authenticate(token).await().indefinitely(); <2>
        ...
    }

}
----
<1> For default tenant, the `TenantFeature` qualifier is optional.
<2> Executes token verification and converts the token to a `SecurityIdentity`.

WARNING: <<tenant-config-resolver>> is currently not supported. Authentication that requires dynamic tenant will fail.

> == References

Reviewer: Hi @michalvavrik Example looks very practical, thanks.

The only I've realized is that I don't think we are modifying the right doc here. This is feature is really not about multi-tenancy at all, we only make sure here the right configuration is picked up for he offline token verification to succeed - but it can be noted in the example.

IMHO the doc updates belong to the `security-oidc-bearer-token-authentication.adoc`, I believe it would be the main flow for such use cases the token has arrived from somewhere, etc. Iif the practice shows it then we can link to the section in the bearer token guide from the code flow guide, but right now I believe it is not necessary.

Please move the docs, IMHO it would make it easier for the users who may need to find this feature in the bearer token concept doc
